zcash_client_backend/src/encoding.rs

```diff
@@ -26,15 +26,55 @@ where
     bech32::encode(hrp, data.to_base32(), Variant::Bech32).expect("hrp is invalid")
 }
 
-fn bech32_decode<T, F>(hrp: &str, s: &str, read: F) -> Result<Option<T>, Error>
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum KeyError {
+    Bech32Error(Error),
+    IncorrectVariant(Variant),
+    ReadError,
```

Comment:
@nuttycom notes that this also was hard to name. The fact that our `read` functions return `Option<T>` is also eliding parsing error information (somewhat intentionally), but that also suggests that this error is not _quite_ `Bech32DecodeError`, in that the errors are not just from decoding Bech32. It is an error for decoding things that are expected to be encoded as Bech32. For now though, `Bech32DecodeError` is fine as a name for this, and we can revisit this error stack when we refactor the parsing logic.